### PR TITLE
Replace internal tracing hooks with otelsql

### PR DESCRIPTION
**Pull request by jhchabran**
Replace our in-house tracing for sql integration so it uses `otelsql` instead. We're getting more details and tracing on the driver level. 

Notes: 

- I had to patch the library, as it caches too eagerly the trace provider, which result in never having it available in time as we always start the db before the rest in our `main()` functions. 
  - See https://github.com/sourcegraph/otelsql/pull/1 
- I also made a patch to the library so keep feature parity, i.e have the query arguments added as attributes on the span itself. 
  - See https://github.com/sourcegraph/otelsql/pull/2  
  - The arguments are now added as `db.args.$n: value`. 
- [x] `sqlhooks` probably needs to be patched, as it's missing one of the few optional methods and using it with `otelsql.Open` yields some issues. That being said it's highly probable that we end up not needing it at all because otelsql would handle everything.
  - Fixed by wrapping the driver.
  - Once we have looked at the metrics being exported, we may possibly remove entirely the sqlhooks. For the time being, it's a good first iteration as it is (though it comes with a bit a juggling around interfaces). 

## Test plan

Tested locally, see https://github.com/sourcegraph/sourcegraph/issues/37905#issuecomment-1192557027 

Update: added the arguments as attributes: 

CleanShot 2022-08-23 at 17 21 44@2x

### CODE REVIEW

File: otelsql.patch [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+-	Tracer         trace.Tracer
++	Tracer         func() trace.Tracer
+ 
+ 	MeterProvider metric.MeterProvider
+ 	Meter         metric.Meter
+@@ -118,10 +118,12 @@ func newConfig(options ...Option) config {
+ 		opt.Apply(&cfg)
+ 	}
+ 
+-	cfg.Tracer = cfg.TracerProvider.Tracer(
```
Comment (jhchabran) [inline otelsql.patch:18]: This is where it get cached.

**Comment by efritz** [comment]
@jhchabran What is the benefit of this over what we have? AFAIK we already send query text to traces.

**Comment by jhchabran** [comment]
> What is the benefit of this over what we have? AFAIK we already send query text to traces.

@efritz We'd get the tracing on the driver interface methods basically. Is that something that you think would be useful?

**Comment by efritz** [comment]
Oh okay I clicked the link 🤦. This is pretty awesome. I'd like to have this prior to us trying to adopt a connection pooler (like pgbouncer or something) so we can track the connection wait times. Is this something we could graph over time via Prometheus as well?

**Comment by jhchabran** [comment]
Nice! 

> Is this something we could graph over time via Prometheus as well?

@efritz yes, that package supports metrics as well, so we could export that :) 

See https://github.com/XSAM/otelsql for more info!

**Comment by Strum355** [comment]
One thing I will comment is that this seems to add a lot of noise to traces (looking at reset_session specifically) that could be tidier as span events/span logs instead imo, or even exclude certain ones
> We'd get the tracing on the driver interface methods basically
which methods is this in reference to btw? Im not super familiar with it

**Comment by jhchabran** [comment]
@Strum355 yeah, it's configurable so we could adjust that.

File: internal/database/dbconn/open.go [hunk 2]
```diff
@@ -69,27 +73,87 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
 func open(cfg *pgx.ConnConfig) (*sql.DB, error) {
 	registerOnce.Do(registerPostgresProxy)
 
-	db, err := sql.Open("postgres-proxy", stdlib.RegisterConnConfig(cfg))
+	db, err := otelsql.Open(
+		"postgres-proxy",
+		stdlib.RegisterConnConfig(cfg),
+		otelsql.WithTracerProvider(otel.GetTracerProvider()),
+		otelsql.WithSpanOptions(otelsql.SpanOptions{
+			OmitConnResetSession: true,
```
Comment (jhchabran) [inline internal/database/dbconn/open.go:154]: Ignore the reset session spans.

File: internal/database/dbconn/open.go [hunk 3]
```diff
@@ -69,27 +73,87 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
+// otelsql.Conn expects to be implemented.
+func (d *extendedDriver) Open(str string) (driver.Conn, error) {
+	if _, ok := d.Driver.(*sqlhooks.Driver); !ok {
+		return nil, errors.New("sql driver is not a sqlhooks.Driver, aborting")
+	}
+	c, err := d.Driver.Open(str)
+	if err != nil {
+		return nil, err
+	}
+	return &extendedConn{
```
Comment (jhchabran) [inline internal/database/dbconn/open.go:103]: TODO just check if those casts are working and do it like 96

File: go.mod [hunk 4]
```diff
@@ -460,6 +465,8 @@ replace (
 // =================================
 // These entries indicate replace directives that are defined for unknown reasons.
 replace (
+	// Forked until PR is merged upstream TODO @jhchabran
+	github.com/XSAM/otelsql => github.com/sourcegraph/otelsql v0.0.0-20220823190616-afb3dd6ab55c
```
Comment (bobheadxi) [inline go.mod:469]: There is a "Temporary replace directives" section above, let's move this there!

File: internal/database/batch/batch.go [hunk 5]
```diff
@@ -7,9 +7,9 @@ import (
 	"sync"
 
 	"github.com/opentracing/opentracing-go/log"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbconn"
 
 	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
-	"github.com/sourcegraph/sourcegraph/internal/database/dbconn"
```
Comment (bobheadxi) [inline internal/database/batch/batch.go:12]: Nit: I like the original location here better 😛 also the rest of this file looks unmodified

File: internal/database/dbconn/bulk_insertion.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"context"
+)
+
+type key struct{}
+
+var bulkInsertionKey = key{}
+
+// bulkInsertion returns true if the bulkInsertionKey context value is true.
+func bulkInsertion(ctx context.Context) bool {
```
Comment (bobheadxi) [inline internal/database/dbconn/bulk_insertion.go:12]:
```suggestion
// isBulkInsertion indicates if a bulk insertion is occurring within this context.
func isBulkInsertion(ctx context.Context) bool {
```

File: internal/database/dbconn/bulk_insertion.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	v, ok := ctx.Value(bulkInsertionKey).(bool)
+	if !ok {
+		return false
+	}
+	return v
+}
+
+// WithBulkInsertion sets the bulkInsertionKey context value.
+func WithBulkInsertion(ctx context.Context, bulkInsertion bool) context.Context {
```
Comment (bobheadxi) [inline internal/database/dbconn/bulk_insertion.go:21]:
```suggestion
// WithBulkInsertion sets whether or not a bulk insertion is occurring within this context.
func WithBulkInsertion(ctx context.Context, bulkInsertion bool) context.Context {
```

File: internal/database/dbconn/open.go [hunk 8]
```diff
@@ -69,27 +73,115 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
+	driver.ConnPrepareContext
+	driver.ConnBeginTx
+}
+
+// Open returns a conn wrapped through extendedConn, implementing the
+// Ping, ResetSession and CheckNamedValue optional methods that the
+// otelsql.Conn expects to be implemented.
+func (d *extendedDriver) Open(str string) (driver.Conn, error) {
+	if _, ok := d.Driver.(*sqlhooks.Driver); !ok {
+		return nil, errors.New("sql driver is not a sqlhooks.Driver, aborting")
```
Comment (bobheadxi) [inline internal/database/dbconn/open.go:97]:
```suggestion
		return nil, errors.New("sql driver is not a sqlhooks.Driver")
```

File: internal/database/dbconn/open.go [hunk 9]
```diff
@@ -69,27 +73,115 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
+	driver.QueryerContext
+	driver.Conn
+	driver.ConnPrepareContext
+	driver.ConnBeginTx
+}
+
+// Open returns a conn wrapped through extendedConn, implementing the
+// Ping, ResetSession and CheckNamedValue optional methods that the
+// otelsql.Conn expects to be implemented.
+func (d *extendedDriver) Open(str string) (driver.Conn, error) {
```
Comment (bobheadxi) [inline internal/database/dbconn/open.go:95]:
```suggestion
// otelsql.Conn expects to be implemented.
//
// An error is returned if the underlying driver does not implement
// the required interfaces. TODO
func (d *extendedDriver) Open(str string) (driver.Conn, error) {
```

For the TODO here is to document some hints as to scenarios where we might not fulfill all the required interfaces, to help with debugging etc (I'm not sure I fully understand this myself)

File: internal/database/dbconn/open.go [hunk 10]
```diff
@@ -69,27 +73,115 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
+	return n.rawConn().(driver.Pinger).Ping(ctx)
+}
+
+func (n *extendedConn) ResetSession(ctx context.Context) error {
+	return n.rawConn().(driver.SessionResetter).ResetSession(ctx)
+}
+
+func (n *extendedConn) CheckNamedValue(namedValue *driver.NamedValue) error {
+	return n.rawConn().(driver.NamedValueChecker).CheckNamedValue(namedValue)
+}
```
Comment (bobheadxi) [inline internal/database/dbconn/open.go:148]: Common pattern I like - if `extendedConn` should implement `driver.Pinger`, `driver.SessionResetter`, etc. we should "code document" this by making some interface assertions:

```
var _ driver.Pinger = &extendedConn{}
```

etc.

**Review by bobheadxi — APPROVED** [review]
Nice work! Tentative approval, since it seems to work, but have some comments :) I've also reviewed the related PRs to the tracing library

File: internal/database/dbconn/open.go (hunk 10 above)
Comment (jhchabran) [reply internal/database/dbconn/open.go:148]: Yeah, I use it often too, though forgot to see it applies here, and kinda convey nicely the intent 👍

File: internal/database/dbconn/open.go (hunk 9 above)
Comment (jhchabran) [reply internal/database/dbconn/open.go:124]: @bobheadxi I added a diagram to make this clearer https://github.com/sourcegraph/sourcegraph/pull/39300/files#diff-bc34aa163694ecf12848f338a0eccb4a222a5d61dc30061d597fc78f8238c906R79 

I dislike the idea of adding code that is hard to comprehend, it's bad for maintainability. Does the diagram makes it clearer for you?

File: internal/database/dbconn/open.go [hunk 11]
```diff
@@ -69,27 +73,144 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
...
+// └─┬─────┘   └─┬──────────┘   └─┬──────┘     └─┬───────────┘
+//   │           │                │              │
+//   │           │                │              │Implements all SQL driver methods
+//   │           │                │
+//   │           │                │Only implements mandatory ones
+//   │           │                │Ping(), ResetSession() and CheckNamedValue() are missing.
+//   │           │
+//   │           │Implement all SQL driver methods
+//   │
+//   │Expects all SQL driver methods
```
Comment (burmudar) [inline internal/database/dbconn/open.go:99]: Love me some diagrams!

**Review by burmudar — APPROVED** [review]

File: internal/database/dbconn/open.go [hunk 12]
```diff
@@ -69,27 +73,144 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
 	}
 }
 
+// extendedDriver wraps sqlHooks' driver to provide a conn that implements Ping, ResetSession
+// and CheckNamedValue, which is mandatory as otelsql is instrumenting these methods.
```
Comment (burmudar) [inline internal/database/dbconn/open.go:77]:
```suggestion
// and CheckNamedValue, which is mandatory as otelsql is instrumenting these methods.
// For all mandatory methods the sqlHooks driver is used. For the optional methods namely Ping, ResetSession and CheckNamedValue 
// (which the sqlHooks driver does not implement), extendedConn goes to the original default driver.
```
This is more for myself to make sure I understand the diagram here @jhchabran

File: internal/database/dbconn/bulk_insertion.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"context"
+)
+
+type key struct{}
+
+var bulkInsertionKey = key{}
+
+// isBulkInsertion returns true if the bulkInsertionKey context value is true.
```
Comment (bobheadxi) [inline internal/database/dbconn/bulk_insertion.go:11]:
```suggestion
// isBulkInsertion indicates if a bulk insertion is occurring within this context.
```

File: internal/database/dbconn/open.go (hunk 11 above)
Comment (bobheadxi) [reply internal/database/dbconn/open.go:101]: I LOVE THIS DIAGRAM TOO

(how was this made?)

File: internal/database/dbconn/open.go (hunk 9 above)
Comment (bobheadxi) [reply internal/database/dbconn/open.go:126]: LGTM 😁 https://github.com/sourcegraph/sourcegraph/pull/39300/files#r955239797

Comment (jhchabran) [reply internal/database/dbconn/open.go:126]: Answered positively on Slack :)

File: internal/database/dbconn/open.go (hunk 11 above)
Comment (jhchabran) [reply internal/database/dbconn/open.go:101]: @bobheadxi https://asciiflow.com/#/ !

**Pull request merged**